internal/statemachine/state_machine.go

```diff
@@ -0,0 +1,323 @@
+		}
+		stateMachine.startingStep = partialStateMachine.CurrentStep
+	} else {
+		// start from the beginning
+		stateMachine.startingStep = 0
+	}
+	return nil
+}
+
+func (stateMachine *StateMachine) getUntilThruOrdinals() error {
+	// attempt to parse thru and until
+	if stateMachine.Until != "" {
+		// first check if it is a digit
+		if val, err := strconv.Atoi(stateMachine.Until); err == nil {
+			// got a digit, make sure it's within range
+			if val >= 0 && val <= len(stateNames) {
+				stateMachine.UntilOrdinal = val
+			} else {
+				return fmt.Errorf("Provided \"until\" step out of range")
+			}
+		} else if val, exists := stateNames[stateMachine.Until]; exists {
+			stateMachine.UntilOrdinal = val
+		} else {
+			return fmt.Errorf("Invalid value for Until: %s\n", stateMachine.Until)
+		}
+	} else {
+		// go "until" the end
+		stateMachine.UntilOrdinal = len(stateFuncs) + 1
+	}
+
+	if stateMachine.Thru != "" {
+		// first check if it is a digit
+		if val, err := strconv.Atoi(stateMachine.Thru); err == nil {
+			// got a digit, make sure it's within range
+			if val >= 0 && val <= len(stateNames) {
+				stateMachine.ThruOrdinal = val
+			} else {
+				return fmt.Errorf("Provided \"thru\" step out of range")
+			}
+		} else if val, exists := stateNames[stateMachine.Thru]; exists {
+			stateMachine.ThruOrdinal = val
+		} else {
+			return fmt.Errorf("Invalid value for Thru: %s\n", stateMachine.Thru)
+		}
+	} else {
+		// go "thru" the end
+		stateMachine.ThruOrdinal = len(stateFuncs)
+	}
+	return nil
+}
+
+/* For state machine runs that will be resumed,
+ * We need to write the state machine info to disk */
+func (stateMachine *StateMachine) writeMetadata() error {
+	gobfile, err := os.OpenFile(stateMachine.WorkDir+"/ubuntu-image.gob", os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Error opening metadata file for writing: %s\n", stateMachine.WorkDir+"/ubuntu-image.gob")
+	}
+	defer gobfile.Close()
+	enc := gob.NewEncoder(gobfile)
+
+	// no need to check errors, as it will panic if there is one
+	enc.Encode(stateMachine)
+	return nil
+}
+
+/* Clean up workdir. For now this is just deleting the temporary directory if necessary
+ * but will have more functionality added to it later */
+func (stateMachine *StateMachine) cleanup() error {
+	if err := os.RemoveAll(stateMachine.WorkDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Step 0: generate work directory file structure
+func (stateMachine *StateMachine) makeTemporaryDirectories() error {
+	if stateMachine.Debug {
+		fmt.Println("[ 0] make_temporary_directories")
+	}
+
+	// if no workdir was specified, open a /tmp dir
+	if stateMachine.WorkDir == "" {
+		workDir, err := os.MkdirTemp(stateMachine.tempLocation, "ubuntu-image-")
+		if err != nil {
+			return fmt.Errorf("Failed to create temporary directory")
+		}
+		stateMachine.WorkDir = workDir
+		stateMachine.CleanWorkDir = true
+	} else {
+		err := os.Mkdir(stateMachine.WorkDir, 0755)
```

Comment:
MkdirAll?  Do we know for certain that the parents will exist?

Reply:
I checked the behavior of the existing tool and it does indeed create the parent directories if they don't exist, so I will implement this change.